File: android/app/src/main/kotlin/com/fyrbyadditive/smartflasher/protocol/slip/SlipCodec.kt
package com.fyrbyadditive.smartflasher.protocol.slip

/**
 * SLIP (Serial Line Internet Protocol) encoder/decoder
 * Used for framing ESP32 bootloader packets
 */
object SlipCodec {
    // SLIP special characters
    const val FRAME_END: Byte = 0xC0.toByte()
    const val FRAME_ESCAPE: Byte = 0xDB.toByte()
    const val ESCAPED_END: Byte = 0xDC.toByte()      // 0xC0 -> 0xDB 0xDC
    const val ESCAPED_ESCAPE: Byte = 0xDD.toByte()   // 0xDB -> 0xDB 0xDD

    /**
     * Encode data with SLIP framing
     * @param data Raw data to encode
     * @return SLIP-encoded packet with 0xC0 delimiters
     */
    fun encode(data: ByteArray): ByteArray {
        val encoded = mutableListOf<Byte>()

        encoded.add(FRAME_END)

        for (byte in data) {
            when (byte) {
                FRAME_END -> {
                    encoded.add(FRAME_ESCAPE)
                    encoded.add(ESCAPED_END)
                }
                FRAME_ESCAPE -> {
                    encoded.add(FRAME_ESCAPE)
                    encoded.add(ESCAPED_ESCAPE)
                }
                else -> encoded.add(byte)
            }
        }

        encoded.add(FRAME_END)
        return encoded.toByteArray()
    }

    /**
     * Decode a SLIP-framed packet
     * @param slipPacket SLIP-encoded packet (including delimiters)
     * @return Decoded raw data, or null if invalid
     */
    fun decode(slipPacket: ByteArray): ByteArray? {
        val decoded = mutableListOf<Byte>()
        var inEscape = false
        var started = false

        for (byte in slipPacket) {
            if (byte == FRAME_END) {
                if (started && decoded.isNotEmpty()) {
                    return decoded.toByteArray()
                }
                started = true
                decoded.clear()
                continue
            }

            if (!started) continue

            if (inEscape) {
                when (byte) {
                    ESCAPED_END -> decoded.add(FRAME_END)
                    ESCAPED_ESCAPE -> decoded.add(FRAME_ESCAPE)
                    else -> decoded.add(byte) // Invalid escape sequence
                }
                inEscape = false
            } else if (byte == FRAME_ESCAPE) {
                inEscape = true
            } else {
                decoded.add(byte)
            }
        }

        return if (decoded.isEmpty()) null else decoded.toByteArray()
    }
}

/**
 * Stateful SLIP decoder for streaming data
 */
class SlipDecoder {
    private val buffer = mutableListOf<Byte>()
    private var inEscape = false
    private var packetStarted = false

    /**
     * Process incoming bytes and return complete packets
     * @param byte Single byte to process
     * @return Complete decoded packet if one was received, null otherwise
     */
    fun process(byte: Byte): ByteArray? {
        return when (byte) {
            SlipCodec.FRAME_END -> {
                if (packetStarted && buffer.isNotEmpty()) {
                    val packet = buffer.toByteArray()
                    reset()
                    packet
                } else {
                    packetStarted = true
                    buffer.clear()
                    null
                }
            }
            SlipCodec.FRAME_ESCAPE -> {
                if (packetStarted) {
                    inEscape = true
                }
                null
            }
            else -> {
                if (!packetStarted) return null

                if (inEscape) {
                    inEscape = false
                    when (byte) {
                        SlipCodec.ESCAPED_END -> buffer.add(SlipCodec.FRAME_END)
                        SlipCodec.ESCAPED_ESCAPE -> buffer.add(SlipCodec.FRAME_ESCAPE)
                        else -> buffer.add(byte)
                    }
                } else {
                    buffer.add(byte)
                }
                null
            }
        }
    }

    /**
     * Process multiple bytes at once
     * @param data Data to process
     * @return List of complete decoded packets
     */
    fun process(data: ByteArray): List<ByteArray> {
        val packets = mutableListOf<ByteArray>()
        for (byte in data) {
            process(byte)?.let { packets.add(it) }
        }
        return packets
    }

    /**
     * Reset the decoder state
     */
    fun reset() {
        buffer.clear()
        inEscape = false
        packetStarted = false
    }
}

File: linux/src/protocol/SLIPCodec.cpp
// FAME Smart Flasher - Linux Qt Port
// Copyright 2025 Fyrby Additive Manufacturing & Engineering
// SPDX-License-Identifier: Proprietary

#include "SLIPCodec.h"

namespace SLIPCodec {

QByteArray encode(const QByteArray& data)
{
    QByteArray encoded;
    encoded.reserve(data.size() * 2 + 2);

    encoded.append(static_cast<char>(FRAME_END));

    for (int i = 0; i < data.size(); ++i) {
        uint8_t byte = static_cast<uint8_t>(data[i]);
        switch (byte) {
        case FRAME_END:
            encoded.append(static_cast<char>(FRAME_ESCAPE));
            encoded.append(static_cast<char>(ESCAPED_END));
            break;
        case FRAME_ESCAPE:
            encoded.append(static_cast<char>(FRAME_ESCAPE));
            encoded.append(static_cast<char>(ESCAPED_ESCAPE));
            break;
        default:
            encoded.append(static_cast<char>(byte));
            break;
        }
    }

    encoded.append(static_cast<char>(FRAME_END));
    return encoded;
}

QByteArray decode(const QByteArray& slipPacket)
{
    QByteArray decoded;
    decoded.reserve(slipPacket.size());

    bool inEscape = false;
    bool started = false;

    for (int i = 0; i < slipPacket.size(); ++i) {
        uint8_t byte = static_cast<uint8_t>(slipPacket[i]);

        if (byte == FRAME_END) {
            if (started && !decoded.isEmpty()) {
                return decoded;
            }
            started = true;
            decoded.clear();
            continue;
        }

        if (!started) {
            continue;
        }

        if (inEscape) {
            switch (byte) {
            case ESCAPED_END:
                decoded.append(static_cast<char>(FRAME_END));
                break;
            case ESCAPED_ESCAPE:
                decoded.append(static_cast<char>(FRAME_ESCAPE));
                break;
            default:
                // Invalid escape sequence
                decoded.append(static_cast<char>(byte));
                break;
            }
            inEscape = false;
        } else if (byte == FRAME_ESCAPE) {
            inEscape = true;
        } else {
            decoded.append(static_cast<char>(byte));
        }
    }

    return decoded.isEmpty() ? QByteArray() : decoded;
}

} // namespace SLIPCodec

std::vector<QByteArray> SLIPDecoder::process(const QByteArray& data)
{
    std::vector<QByteArray> packets;
    for (int i = 0; i < data.size(); ++i) {
        QByteArray packet = processByte(static_cast<uint8_t>(data[i]));
        if (!packet.isEmpty()) {
            packets.push_back(packet);
        }
    }
    return packets;
}

QByteArray SLIPDecoder::processByte(uint8_t byte)
{
    switch (byte) {
    case SLIPCodec::FRAME_END:
        if (m_packetStarted && !m_buffer.isEmpty()) {
            QByteArray packet = m_buffer;
            reset();
            return packet;
        }
        m_packetStarted = true;
        m_buffer.clear();
        return QByteArray();

    case SLIPCodec::FRAME_ESCAPE:
        if (!m_packetStarted) {
            return QByteArray();
        }
        m_inEscape = true;
        return QByteArray();

    default:
        if (!m_packetStarted) {
            return QByteArray();
        }

        if (m_inEscape) {
            m_inEscape = false;
            switch (byte) {
            case SLIPCodec::ESCAPED_END:
                m_buffer.append(static_cast<char>(SLIPCodec::FRAME_END));
                break;
            case SLIPCodec::ESCAPED_ESCAPE:
                m_buffer.append(static_cast<char>(SLIPCodec::FRAME_ESCAPE));
                break;
            default:
                m_buffer.append(static_cast<char>(byte));
                break;
            }
        } else {
            m_buffer.append(static_cast<char>(byte));
        }
        return QByteArray();
    }
}

void SLIPDecoder::reset()
{
    m_buffer.clear();
    m_inEscape = false;
    m_packetStarted = false;
}
